cli/sql: new options to help troubleshooting

knz opened the pull request:
Fixes #18120.

cc @cockroachdb/sql-execution 

See the two commits for details.

knz committed:
cli: add a new command-line parameter --echo-sql

This new parameter, when set, causes all SQL strings sent to the
server to also be printed on the client's standard error stream.

This is intended for troubleshooting and testing.

This parameter can also be enabled/disabled dynamically in the CLI
shell using `\set echo` / `\unset echo`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18127)
<!-- Reviewable:end -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Thanks a lot!

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/cli/sql.go, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127#-Ksy4v4FHfmS20h5IU_e:-Ksy4v4FHfmS20h5IU_f:b-m1xz1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c845132bd3db00b89f74fe01c0fff0b3a74c03d/pkg/cli/sql.go#L81)):*
> ```Go
> 	// properly.
> 	normalizeHistory bool
> 	// smartPrompt indicates whether to update the prompt using queries to the server.
> ```

nit: give a hint as to what the queries and the prompt updates would be

---

*[pkg/cli/interactive_tests/test_local_cmds.tcl, line 67 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127#-Ksy4mlHMBNTJvtu0yEd:-Ksy4mlHMBNTJvtu0yEe:b-rwdo6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb58b1dd22f24372f43ea7d135e2ffdbaf0049b/pkg/cli/interactive_tests/test_local_cmds.tcl#L67)):*
> ```Tcl
> start_test "Check that \\set without argument prints the current options"
> send "\\set\r"
> eexpect "5 rows"
> ```

how come this changed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127#-:-Ksy50r0CrQLjgv9ZCmO:b-945yy1)*
<!-- Sent from Reviewable.io -->

knz committed:
cli/sql: new option smart_prompt to control server queries to update the prompt

Requested by @andreimatei; this change is intended to help testing and
troubleshooting.

This patch introduces the client-side option `smart_prompt`. When
enabled:

- the CLI shell sends requests to the server in between inputs to
  update the prompt. Currently this includes a request for the
  database name and another for the current transaction status.

- the CLI shell suggests the user to continue the input across
  multiple lines if the first line contains a BEGIN statement but not
  also a COMMIT, so as to encourage sending the entire transaction in
  one pgwire request.

This option is enabled by default in interactive shells. It can be
controlled as usual with `\set` and `\unset`.

knz commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/cli/sql.go, line 81 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127#-Ksy4v4FHfmS20h5IU_e:-KsyfHS93NqBX7nE91ma:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c845132bd3db00b89f74fe01c0fff0b3a74c03d/pkg/cli/sql.go#L81)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: give a hint as to what the queries and the prompt updates would be
</blockquote></details>

Done.

---

*[pkg/cli/interactive_tests/test_local_cmds.tcl, line 67 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127#-Ksy4mlHMBNTJvtu0yEd:-KsyfIv7PsdJJcH2yu56:bp4g551) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb58b1dd22f24372f43ea7d135e2ffdbaf0049b/pkg/cli/interactive_tests/test_local_cmds.tcl#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come this changed?
</blockquote></details>

The test doesn't need to check how many options there are in total, just that they are printed. I figured it's silly to have to update this counter here every time we add a new option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18127)*
<!-- Sent from Reviewable.io -->
